### ISSUE TITLE
alpine:init issue

### ISSUE THREAD

**Issue by TheSteveBurgess** [EmilStenstrom/django-components#818]
I've got a super weird issue and i was wondering if anyone else had encountered it?

I have a number of Django Components that create and register their Alpine.js x-data in the Django Components js file. 

Something like this:

```
 document.addEventListener('alpine:init', () => {
        Alpine.data('somexdata', () => ({
            somevalue: false
        }))
    })
```

This works fine when I'm referencing Alpine.js externally via its CDN link. However, if I’m referencing Alpine locally then the Alpine.data method is never called. It’s like the alpine:init method has already fired before the code that subscribes to the alpine:init event is added to the page

Normally you would need to add this event hook after the Alpine.js script is loaded but before any Alpine components are initialised. 

In the previous version of Django Components I was using (0.73 i think?) I could call component_js_dependencies and have the Django Component scripts added into the head of the page. Using component_js_dependencies to place the scripts in the head now results in a JavaScript error. If i put component_js_dependencies at the top of the body it’s still fails. I'm thinking this is because rather than writing script tags there is now a JS loader that loads the scripts and the scripts are now added to the page after the apline:init event has fired. 

Is there a way to have component_js_dependencies just out put script tags rather than use the JS script loader?

**Comment by JuroOravec** [comment]
@TheSteveBurgess Did you experience this on v0.110 or later? I think this might be the same issue as https://github.com/EmilStenstrom/django-components/issues/791.

- What do you mean by "referencing Alpine locally"? Do you by that mean the snippet that you've included?

- How do you load AlpineJS? Do you specify the CND link explicitly in the HTML, or on Component as `Component.Media.js`?

- That snippet, where is it defined? Is it defined on a Component? As `Component.js` or `Component.Media.js`?

For now I suggest to update only to v0.102, which is the latest release before v0.110. The issue is that all component JS scripts are loaded simultaneously, so your snippet may or may not trigger, depending on whether it's loaded before or after the core AlpineJS library.

Instead, what I want to implement as part of https://github.com/EmilStenstrom/django-components/issues/791 is to load the scripts serially. So first the scripts in `Component.Media.js` would be loaded one-by-one, and only after that the script in `Component.js` would be loaded.

---

Although I'm not sure what to do with the loading of AlpineJS.

Because if we define the CDN script it like so:

```html
<script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>
```

Then even with `defer` tag, it might run the AlpineJS code before the component's JS is all loaded and executed. So you might still end up with some JS scripts where ` document.addEventListener('alpine:init', ...)` would NOT run.

---

Actually, a workaround might be to load AlpineJS as an JS / ECMAScript module (`.esm`), so that we have control over WHEN it is being imported.

1. Create a component that defines only the AlpineJS scripts. And it imports them as modules:
	```py
    # Same as original `Media`, except `Media.render_js()`
    # sets `type="module"`
    class JSModuleMedia(Media):
        def render_js(self):
            tags = []
            for path in self._js:
                if hasattr(path, "__html__"):
                    tag = path.__html__()
                else:
                    tag = format_html(
                        '<script type="module" src="{}"></script>',
                        self.absolute_path(path)
                    )
            return tags

    @register("alpine")
    class Alpine(Component):
        media_class = JSModuleMedia
        js = """
          import Alpine from 'https://cdn.jsdelivr.net/npm/alpinejs@3.14.6/dist/module.esm.min.js'
 
          window.Alpine = Alpine;
          Alpine.start();
       """
	```

2. Insert the Alpine component at the end of the body of the HTML.

    So if you have some root or layout component that defines the HTML, you'd insert it AFTER everything else:
    ```django
    <html>
      <head>
        ...
      </head>
      <body>
        {% slot "content" default  / %}

        {# INSERT ALPINE HERE #}
        {% component "alpine" / %}
      </body>
    </html>
    ```

BUT! Even this workaround would work only once https://github.com/EmilStenstrom/django-components/issues/791 is fixed, to ensure that the `Component.js` of the Alpine component would run as last.

---

A proper fix would be to have an AlpineJS plugin, but that's about a month away still.

**Comment by TheSteveBurgess** [comment]
Thanks for the super fast response :)

The version i was testing it on is 0.114.

When i say "referencing Alpine locally" I mean pointing the script tag to a static file within Django so something like this

```
<script defer src="/static/alpine.js"></script>
```

The snippet is being defined in a js file for the component and then added via the Media class.

I'll try reverting to v0.102 and see what happens.

I think loading Alpine as a module and then being able to control the order of script execution (and then initialising Alpine after all scripts have loaded) could work . I do wonder how that would work with existing Alpine code already in the page that isn't part of a component though? 

I tried something similar using JSPath and adding Alpine directly as a Media script on the component with defer added to the script tag. This did work, however i had issues with other Alpine on the page intermittently failing to load.

I wonder if a fix (at least for the Alpine specific apline:init issue) would be to allow a user to specify that a script is to be loaded at the end of the head tag and that it should be written into the page directly and not loaded via the JS loader. There would have to be some caveats with that though as I'm guessing that wouldn't play well with html fragment loading.

**Comment by TheSteveBurgess** [comment]
I think that should work great 😊

**Comment by JuroOravec** [comment]
Good news, the proof of concept for the above works 🎉 

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8" />
    <script defer src="https://unpkg.com/alpinejs"></script>
  </head>

  <body>
    <div>test</div>

    <div x-data="gencomp">
      GEN_COMP: 
      <div x-text="somevalue"></div>
    </div>

    <script>
      debugger;
      const parseScriptTag = (tag) => {
        const scriptNode = new DOMParser().parseFromString(tag, 'text/html').querySelector('script');
        return scriptNode;
      };

      const cloneNode = (srcNode) => {
        const targetNode = document.createElement(srcNode.tagName);
        for (const attr of srcNode.attributes) {
          targetNode.setAttributeNode(attr.cloneNode());
        }
        targetNode.innerHTML = srcNode.innerHTML;
        return targetNode;
      };

      console.log('I SHOULD BE CALLED FIRST');

      const selfEl = document.currentScript;

      const scriptTag = parseScriptTag(
        '<' + 'script>' + `
          debugger;
          console.log('I SHOULD BE CALLED SECOND');
          console.log(document.currentScript);

          document.addEventListener('alpine:init', () => {
            Alpine.data('gencomp', () => ({
                somevalue: 123,
            }))
         })
      ` + '</' + 'script>'
      );
      const scripTagCopy = cloneNode(scriptTag);
      selfEl.parentNode.insertBefore(scripTagCopy, selfEl.nextSibling);

      // See https://developer.mozilla.org/en-US/docs/Web/API/Document/currentScript
      // selfEl.remove();
    </script>

    <script>
      debugger;
      console.log('I SHOULD BE CALLED LAST');
    </script>
  </body>
</html>

```

**Comment by JuroOravec** [comment]
Sorry for spamming, I'm just caring too much about this issue, because if this didn't work that could be a dealbreaker for me using django-components at work, as I heavily rely on AlpineJS there.

But I can confirm that I got things working with AlpineJS 🎉 , while still respecting the conditions I outlined above:

>   Thus, I think we should achieve that:
>
>   1. Scripts in `Media.js` are executed in the order they are defined
>
>   2. Scripts in `Media.js` will have access to the variables defined in previous scripts
>
>   3. Scripts in `Media.js` will be still managed by our dependency manager, so they will be inserted and executed ONLY IF THEY WERE NOT LOADED BEFORE.
>
>   4. Script in `Component.js` will be executed AFTER the dependencies in `Media.js` and will have access to their variables.
> 
>   5. ALL scripts will be loaded by being inserted into the HTML. So, assuming they all load synchronously, should mean that they load before other inlined `<script>` tags like the AlpineJS CDN.

**Comment by JuroOravec** [comment]
This was fixed and released in v0.116 🎉 

Although I recommend to wait until v0.117 - there's one more known bug that may break the HTML templates sometimes, see https://github.com/EmilStenstrom/django-components/issues/793
